File: DuAnTauHoaCom/backend/src/services/AuthService.js
const bcrypt = require('bcryptjs')
const jwt = require('jsonwebtoken')
const TaiKhoanRepo = require('../repositories/TaiKhoanRepository')

const SALT_ROUNDS = 10

const register = async ({ email, matKhau, hoTen, soDienThoai }) => {
  const exists = await TaiKhoanRepo.emailExists(email)
  if (exists) throw { status: 400, message: 'Email đã được đăng ký' }

  const hashedPwd = await bcrypt.hash(matKhau, SALT_ROUNDS)
  const taiKhoan = await TaiKhoanRepo.create({
    email: email.toLowerCase().trim(),
    mat_khau: hashedPwd,
    ho_ten: hoTen.trim(),
    so_dien_thoai: soDienThoai || null,
    vai_tro: 'khach_hang',
    trang_thai: 'hoat_dong',
  })

  const token = signToken(taiKhoan)
  return { token, user: sanitize(taiKhoan) }
}

const login = async ({ email, matKhau }) => {
  const taiKhoan = await TaiKhoanRepo.findByEmail(email)
  if (!taiKhoan) throw { status: 401, message: 'Email hoặc mật khẩu không đúng' }
  if (taiKhoan.trang_thai !== 'hoat_dong') throw { status: 403, message: 'Tài khoản đã bị khóa' }

  const valid = await bcrypt.compare(matKhau, taiKhoan.mat_khau)
  if (!valid) throw { status: 401, message: 'Email hoặc mật khẩu không đúng' }

  const token = signToken(taiKhoan)
  return { token, user: sanitize(taiKhoan) }
}

const getProfile = async (idTaiKhoan) => {
  const taiKhoan = await TaiKhoanRepo.findById(idTaiKhoan)
  if (!taiKhoan) throw { status: 404, message: 'Không tìm thấy tài khoản' }
  return sanitize(taiKhoan)
}

const updateProfile = async (idTaiKhoan, data) => {
  const allowed = ['ho_ten', 'so_dien_thoai', 'ngay_sinh', 'gioi_tinh']
  const update = {}
  allowed.forEach(k => { if (data[k] !== undefined) update[k] = data[k] })
  await TaiKhoanRepo.update(idTaiKhoan, update, 'id_tai_khoan')
  return getProfile(idTaiKhoan)
}

const changePassword = async (idTaiKhoan, { matKhauCu, matKhauMoi }) => {
  const taiKhoan = await TaiKhoanRepo.findById(idTaiKhoan)
  const valid = await bcrypt.compare(matKhauCu, taiKhoan.mat_khau)
  if (!valid) throw { status: 400, message: 'Mật khẩu cũ không đúng' }
  const hashed = await bcrypt.hash(matKhauMoi, SALT_ROUNDS)
  await TaiKhoanRepo.update(idTaiKhoan, { mat_khau: hashed }, 'id_tai_khoan')
}

const signToken = (taiKhoan) =>
  jwt.sign(
    { id: taiKhoan.id_tai_khoan, email: taiKhoan.email, role: taiKhoan.vai_tro },
    process.env.JWT_SECRET,
    { expiresIn: process.env.JWT_EXPIRES_IN || '365d' }
  )

// Gia hạn token: lấy thông tin từ token cũ và tạo token mới
const refreshToken = async (idTaiKhoan) => {
  const taiKhoan = await TaiKhoanRepo.findById(idTaiKhoan)
  if (!taiKhoan) throw { status: 404, message: 'Không tìm thấy tài khoản' }
  if (taiKhoan.trang_thai !== 'hoat_dong') throw { status: 403, message: 'Tài khoản đã bị khóa' }
  const token = signToken(taiKhoan)
  return { token, user: sanitize(taiKhoan) }
}

const sanitize = (tk) => ({
  id: tk.id_tai_khoan,
  email: tk.email,
  hoTen: tk.ho_ten,
  soDienThoai: tk.so_dien_thoai,
  ngaySinh: tk.ngay_sinh,
  gioiTinh: tk.gioi_tinh,
  vaiTro: tk.vai_tro,
  ngayTao: tk.ngay_tao,
})

module.exports = { register, login, getProfile, updateProfile, changePassword, refreshToken }

File: java-backend/src/main/java/com/klntrain/service/AuthService.java
package com.klntrain.service;

import com.klntrain.dto.request.LoginRequest;
import com.klntrain.dto.request.RegisterRequest;
import com.klntrain.entity.TaiKhoan;
import com.klntrain.repository.TaiKhoanRepository;
import com.klntrain.security.JwtTokenProvider;
import com.klntrain.util.AppException;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.Map;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final TaiKhoanRepository taiKhoanRepo;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider tokenProvider;

    public Map<String, Object> register(RegisterRequest req) {
        if (taiKhoanRepo.existsByEmail(req.getEmail().toLowerCase())) {
            throw AppException.badRequest("Email đã được đăng ký");
        }

        TaiKhoan tk = TaiKhoan.builder()
            .email(req.getEmail().toLowerCase().strip())
            .matKhau(passwordEncoder.encode(req.getMatKhau()))
            .hoTen(req.getHoTen().strip())
            .soDienThoai(req.getSoDienThoai())
            .vaiTro("khach_hang")
            .trangThai("hoat_dong")
            .build();

        tk = taiKhoanRepo.save(tk);
        String token = tokenProvider.generateToken(tk.getIdTaiKhoan(), tk.getEmail(), tk.getVaiTro());
        return Map.of("token", token, "user", sanitize(tk));
    }

    public Map<String, Object> login(LoginRequest req) {
        TaiKhoan tk = taiKhoanRepo.findByEmail(req.getEmail().toLowerCase())
            .orElseThrow(() -> AppException.unauthorized("Email hoặc mật khẩu không đúng"));

        if (!"hoat_dong".equals(tk.getTrangThai())) {
            throw AppException.unauthorized("Tài khoản đã bị khóa");
        }
        if (!passwordEncoder.matches(req.getMatKhau(), tk.getMatKhau())) {
            throw AppException.unauthorized("Email hoặc mật khẩu không đúng");
        }

        String token = tokenProvider.generateToken(tk.getIdTaiKhoan(), tk.getEmail(), tk.getVaiTro());
        return Map.of("token", token, "user", sanitize(tk));
    }

    public Map<String, Object> getProfile(Long idTaiKhoan) {
        TaiKhoan tk = taiKhoanRepo.findById(idTaiKhoan)
            .orElseThrow(() -> AppException.notFound("Không tìm thấy tài khoản"));
        return sanitize(tk);
    }

    public Map<String, Object> updateProfile(Long idTaiKhoan, Map<String, Object> data) {
        TaiKhoan tk = taiKhoanRepo.findById(idTaiKhoan)
            .orElseThrow(() -> AppException.notFound("Không tìm thấy tài khoản"));

        if (data.containsKey("hoTen"))       tk.setHoTen((String) data.get("hoTen"));
        if (data.containsKey("soDienThoai")) tk.setSoDienThoai((String) data.get("soDienThoai"));

        taiKhoanRepo.save(tk);
        return sanitize(tk);
    }

    public void changePassword(Long idTaiKhoan, String matKhauCu, String matKhauMoi) {
        TaiKhoan tk = taiKhoanRepo.findById(idTaiKhoan)
            .orElseThrow(() -> AppException.notFound("Không tìm thấy tài khoản"));

        if (!passwordEncoder.matches(matKhauCu, tk.getMatKhau())) {
            throw AppException.badRequest("Mật khẩu cũ không đúng");
        }
        tk.setMatKhau(passwordEncoder.encode(matKhauMoi));
        taiKhoanRepo.save(tk);
    }

    public Map<String, Object> refreshToken(Long idTaiKhoan) {
        TaiKhoan tk = taiKhoanRepo.findById(idTaiKhoan)
            .orElseThrow(() -> AppException.notFound("Không tìm thấy tài khoản"));
        if (!"hoat_dong".equals(tk.getTrangThai())) {
            throw AppException.unauthorized("Tài khoản đã bị khóa");
        }
        String token = tokenProvider.generateToken(tk.getIdTaiKhoan(), tk.getEmail(), tk.getVaiTro());
        return Map.of("token", token, "user", sanitize(tk));
    }

    private Map<String, Object> sanitize(TaiKhoan tk) {
        return Map.of(
            "id",          tk.getIdTaiKhoan(),
            "email",       tk.getEmail(),
            "hoTen",       tk.getHoTen(),
            "soDienThoai", tk.getSoDienThoai() != null ? tk.getSoDienThoai() : "",
            "vaiTro",      tk.getVaiTro(),
            "trangThai",   tk.getTrangThai()
        );
    }
}
